File: Tutorials/Codes/System_Dynamics_Course/Chapter20/Lesson1/Chapter20_Lesson1.cpp
// Chapter20_Lesson1.cpp
/*
System Dynamics — Chapter 20, Lesson 1
Nonlinear Maps and Continuous-Time Chaotic Systems (Logistic Map, Lorenz System)

This program:
1) Simulates the logistic map and writes x_n to CSV.
2) Integrates the Lorenz system using classical RK4 and writes trajectory to CSV.

Build (example):
  g++ -O2 -std=c++17 Chapter20_Lesson1.cpp -o ch20_l1

Outputs:
  logistic.csv  (n, x_n)
  lorenz.csv    (t, x, y, z)
*/

#include <cmath>
#include <fstream>
#include <iostream>
#include <vector>

static std::vector<double> logistic_map(double r, double x0, int n) {
    std::vector<double> x(n + 1);
    x[0] = x0;
    for (int k = 0; k < n; ++k) {
        x[k + 1] = r * x[k] * (1.0 - x[k]);
    }
    return x;
}

struct Vec3 {
    double x, y, z;
    Vec3() : x(0), y(0), z(0) {}
    Vec3(double x_, double y_, double z_) : x(x_), y(y_), z(z_) {}
};

static Vec3 operator+(const Vec3& a, const Vec3& b) { return Vec3(a.x + b.x, a.y + b.y, a.z + b.z); }
static Vec3 operator-(const Vec3& a, const Vec3& b) { return Vec3(a.x - b.x, a.y - b.y, a.z - b.z); }
static Vec3 operator*(double s, const Vec3& a) { return Vec3(s * a.x, s * a.y, s * a.z); }
static Vec3 operator*(const Vec3& a, double s) { return s * a; }

static Vec3 lorenz_rhs(double /*t*/, const Vec3& s, double sigma, double rho, double beta) {
    Vec3 ds;
    ds.x = sigma * (s.y - s.x);
    ds.y = s.x * (rho - s.z) - s.y;
    ds.z = s.x * s.y - beta * s.z;
    return ds;
}

static Vec3 rk4_step(double t, const Vec3& x, double h, double sigma, double rho, double beta) {
    Vec3 k1 = lorenz_rhs(t, x, sigma, rho, beta);
    Vec3 k2 = lorenz_rhs(t + 0.5 * h, x + 0.5 * h * k1, sigma, rho, beta);
    Vec3 k3 = lorenz_rhs(t + 0.5 * h, x + 0.5 * h * k2, sigma, rho, beta);
    Vec3 k4 = lorenz_rhs(t + h, x + h * k3, sigma, rho, beta);
    return x + (h / 6.0) * (k1 + 2.0 * k2 + 2.0 * k3 + k4);
}

int main() {
    // Logistic map
    const double r = 3.8;
    const double x0 = 0.2;
    const int n = 2000;
    auto x = logistic_map(r, x0, n);

    std::ofstream flog("logistic.csv");
    flog << "n,x\n";
    for (int k = 0; k <= n; ++k) {
        flog << k << "," << x[k] << "\n";
    }
    flog.close();

    // Lorenz RK4
    const double sigma = 10.0;
    const double rho = 28.0;
    const double beta = 8.0 / 3.0;
    const double t0 = 0.0;
    const double tf = 40.0;
    const double h = 0.005;

    int steps = static_cast<int>(std::ceil((tf - t0) / h));
    std::ofstream flor("lorenz.csv");
    flor << "t,x,y,z\n";

    double t = t0;
    Vec3 s(1.0, 1.0, 1.0);
    flor << t << "," << s.x << "," << s.y << "," << s.z << "\n";
    for (int k = 0; k < steps; ++k) {
        s = rk4_step(t, s, h, sigma, rho, beta);
        t += h;
        flor << t << "," << s.x << "," << s.y << "," << s.z << "\n";
    }
    flor.close();

    std::cout << "Wrote logistic.csv and lorenz.csv\n";
    return 0;
}

File: Tutorials/Codes/System_Dynamics_Course/Chapter20/Lesson1/Chapter20_Lesson1.java
// Chapter20_Lesson1.java
/*
System Dynamics — Chapter 20, Lesson 1
Nonlinear Maps and Continuous-Time Chaotic Systems (Logistic Map, Lorenz System)

This program:
1) Simulates logistic map and writes logistic.csv.
2) Integrates Lorenz system via RK4 and writes lorenz.csv.

Compile & run:
  javac Chapter20_Lesson1.java
  java Chapter20_Lesson1

Outputs:
  logistic.csv  (n, x_n)
  lorenz.csv    (t, x, y, z)
*/

import java.io.BufferedWriter;
import java.io.FileWriter;
import java.io.IOException;

public class Chapter20_Lesson1 {

    // Logistic map x_{n+1} = r x_n (1 - x_n)
    static double[] logisticMap(double r, double x0, int n) {
        double[] x = new double[n + 1];
        x[0] = x0;
        for (int k = 0; k < n; k++) {
            x[k + 1] = r * x[k] * (1.0 - x[k]);
        }
        return x;
    }

    static class Vec3 {
        double x, y, z;
        Vec3(double x, double y, double z) { this.x = x; this.y = y; this.z = z; }
        Vec3 add(Vec3 b) { return new Vec3(this.x + b.x, this.y + b.y, this.z + b.z); }
        Vec3 mul(double s) { return new Vec3(s * this.x, s * this.y, s * this.z); }
    }

    static Vec3 lorenzRHS(double t, Vec3 s, double sigma, double rho, double beta) {
        double dx = sigma * (s.y - s.x);
        double dy = s.x * (rho - s.z) - s.y;
        double dz = s.x * s.y - beta * s.z;
        return new Vec3(dx, dy, dz);
    }

    static Vec3 rk4Step(double t, Vec3 x, double h, double sigma, double rho, double beta) {
        Vec3 k1 = lorenzRHS(t, x, sigma, rho, beta);
        Vec3 k2 = lorenzRHS(t + 0.5 * h, x.add(k1.mul(0.5 * h)), sigma, rho, beta);
        Vec3 k3 = lorenzRHS(t + 0.5 * h, x.add(k2.mul(0.5 * h)), sigma, rho, beta);
        Vec3 k4 = lorenzRHS(t + h, x.add(k3.mul(h)), sigma, rho, beta);
        return x.add(k1.mul(h / 6.0))
                .add(k2.mul(h / 3.0))
                .add(k3.mul(h / 3.0))
                .add(k4.mul(h / 6.0));
    }

    public static void main(String[] args) throws IOException {
        // Logistic map
        double r = 3.8;
        double x0 = 0.2;
        int n = 2000;
        double[] x = logisticMap(r, x0, n);

        try (BufferedWriter w = new BufferedWriter(new FileWriter("logistic.csv"))) {
            w.write("n,x\n");
            for (int k = 0; k <= n; k++) {
                w.write(k + "," + x[k] + "\n");
            }
        }

        // Lorenz RK4
        double sigma = 10.0;
        double rho = 28.0;
        double beta = 8.0 / 3.0;

        double t0 = 0.0;
        double tf = 40.0;
        double h = 0.005;
        int steps = (int) Math.ceil((tf - t0) / h);

        Vec3 s = new Vec3(1.0, 1.0, 1.0);
        double t = t0;

        try (BufferedWriter w = new BufferedWriter(new FileWriter("lorenz.csv"))) {
            w.write("t,x,y,z\n");
            w.write(t + "," + s.x + "," + s.y + "," + s.z + "\n");
            for (int k = 0; k < steps; k++) {
                s = rk4Step(t, s, h, sigma, rho, beta);
                t += h;
                w.write(t + "," + s.x + "," + s.y + "," + s.z + "\n");
            }
        }

        System.out.println("Wrote logistic.csv and lorenz.csv");
    }
}
